File: locator_py/plot_locator.py
import argparse, scipy as sp, pandas as pd, os, numpy as np, matplotlib as mpl, zarr
from matplotlib import pyplot as plt
from math import sin, cos, sqrt, atan2, radians
from sklearn.neighbors import KernelDensity
from mpl_toolkits.axes_grid1.inset_locator import inset_axes

parser=argparse.ArgumentParser(description='Plot summary of a set of locator predictions.')
parser.add_argument('--infile',help='path to folder with .predlocs files')
parser.add_argument('--sample_data',help='path to sample_data file (should be WGS1984 x/y if map=True')
parser.add_argument('--out',help='path to output (will be appended with _typeofplot.pdf')
parser.add_argument('--width',default=10,type=float,help='width in inches of the output plot. default=5')
parser.add_argument('--height',default=8,type=float,help='height in inches of the output plot. default=4')
parser.add_argument('--samples',default=None,nargs='+',help='samples IDs to plot, separated by spaces. e.g. sample1 sample2 sample3. default=None')
parser.add_argument('--nsamples',default=9,type=int,help='if no --samples argument is provided, --nsamples random samples will be plotted. default=9')
parser.add_argument('--ncol',default=3,type=int,help='number of columns for multipanel plots (should evenly divide --nsamples, otherwise nsamples will supersede). default=3')
parser.add_argument('--error',default=True,action='store_true',help='calculate error and plot summary? requires known locations for all samples. T/F. default=False')
parser.add_argument('--plot',default=False,action='store_true',help='make plots of predictions? default=True')
parser.add_argument('--basemap',default=False,action='store_true',help='plot basemap? default=False')
parser.add_argument('--longlat',default=False,action='store_true',help='set to True if coordinates are x and y in decimal degrees to print error in kilometers. default=False')
parser.add_argument('--silence',default=False,action='store_true',help='no terminal output. T/F. default=True')
parser.add_argument('--training_samples',default=None,help='path to training metadata file for plotting training locations, if provided. default=None')
args=parser.parse_args()

if args.basemap:
    mapp=zarr.open('map.zarr',mode='r')

def kdepred(xcoords,ycoords): # kernel density
    try:
        coords=list(zip(xcoords,ycoords))
        density=KernelDensity(kernel='gaussian',bandwidth=0.2).fit(coords) # bandwidth
        e=density.score_samples(coords)
        max_index=int((np.argwhere(e==np.amax(e)).tolist())[0][0])
        kx=xcoords[max_index]
        ky=ycoords[max_index]
    except Exception:
        print('oops')
        kx=np.mean(xcoords)
        ky=np.mean(ycoords)
    return kx,ky

def centroid(xcoords,ycoords): # geographic centroid
    coords=np.array(list(zip(xcoords,ycoords)))
    length=coords.shape[0]
    sum_x=np.sum(coords[:,0])
    sum_y=np.sum(coords[:,1])
    return sum_x/length, sum_y/length

def distance_km(xpred,ypred,x,y): # distance in km if longlat==True
    dlon=xpred-x
    dlat=ypred-y
    a=sin(dlat/2)**2+cos(y)*cos(ypred)*sin(dlon/2)**2
    c=2*atan2(sqrt(a),sqrt(1-a))
    return 6373.0*c

def distance(xpred,ypred,x,y): # hypotenuse distance otherwise
    xc=xpred-x
    yc=ypred-y
    return np.hypot(xc,yc)

def get_closest(floats,value): # closest float
    diff=(np.abs(floats-value))
    return diff.argmin()

# load data
if not args.silence:
    print('loading data')
if 'predlocs.txt' in args.infile:
    aeg=pd.DataFrame(infile)
else:
    files=[x for x in os.listdir(args.infile)]
    files=[x for x in files if 'predlocs' in x]
    aeg=pd.read_csv(args.infile+'/'+files[0])
    for i in range(len(files[1:])):
        a=pd.read_csv(args.infile+'/'+files[i])
        aeg=aeg.append(a,ignore_index=True)
aeg=aeg.reset_index(drop=True)
aeg=aeg.rename(columns={'x':'xpred','y':'ypred'})
locs=pd.read_csv(args.sample_data,sep='\t')
aeg=pd.merge(aeg,locs,on='sampleID')
samples=aeg.sampleID.unique()
if args.samples:
    samples=args.samples
else:
    samples=np.random.choice(samples,args.nsamples,replace=False)

# calculate error
if args.error:
    if not args.silence:
        print('calculating error')
    # get error for centroids and max kernel density locations
    bp={}
    kd_x=[]
    kd_y=[]
    gc_x=[]
    gc_y=[]
    loc_x=[]
    loc_y=[]
    kd_dists=[]
    gc_dists=[]
    ids=aeg.sampleID.unique()
    for item in ids:
        xcoords=aeg.loc[aeg['sampleID']==item,'xpred'].tolist()
        ycoords=aeg.loc[aeg['sampleID']==item,'ypred'].tolist()
    # actual location
        x=aeg.loc[aeg['sampleID']==item]['x'].tolist()
        y=aeg.loc[aeg['sampleID']==item]['y'].tolist()
        loc_x.append(x[0])
        loc_y.append(y[0])
    # kd centroids
        k=kdepred(xcoords,ycoords)
        kd_x.append(k[0])
        kd_y.append(k[1])
        if args.longlat:
            kd_dists.append(distance_km(k[0],k[1],x[0],y[0]))
        else:
            kd_dists.append(distance(k[0],k[1],x[0],y[0]))
     # gc centroids
        g=centroid(xcoords,ycoords)
        gc_x.append(g[0])
        gc_y.append(g[1])
        if args.longlat:
            gc_dists.append(distance_km(g[0],g[1],x[0],y[0]))
        else:
            gc_dists.append(distance(g[0],g[1],x[0],y[0]))
    # save to tsv
    bp.update({'sampleID':ids,'x':loc_x,'y':loc_y,'kd_x':kd_x,'kd_y':kd_y,'gc_x':gc_x,'gc_y':gc_y})
    bp=pd.DataFrame(bp)
    bp.to_csv(args.out+'_centroids.txt',index=False,sep='\t')
    # print results
    if not args.silence:
        print('mean kernel peak error = ' + str(np.mean(kd_dists)))
        print('median kernel peak error = ' + str(np.median(kd_dists)))
        print('90% CI for kernel peak error = ' + str(np.quantile(kd_dists,0.05))+' '+str(np.quantile(kd_dists,0.95)))
        print('mean centroid error = ' + str(np.mean(gc_dists)))
        print('median centroid error = ' + str(np.median(gc_dists)))
        print('90% CI for centroid error = ' + str(np.quantile(gc_dists,0.05))+' '+str(np.quantile(gc_dists,0.95)))
if not args.plot:
    quit()
# plot
if args.plot:
    if args.training_samples: # get training samples
        training_samples=pd.read_csv(args.training_samples,sep='\t')
        ts_x=(training_samples['x'][~np.isnan(training_samples['x'].to_numpy())])
        ts_y=(training_samples['y'][~np.isnan(training_samples['y'].to_numpy())])
        print(ts_x,ts_y)
    if not args.silence:
        print('plotting')
    if args.samples:
        s=len(args.samples)
    else:
        s=args.nsamples
    r=s/args.ncol
    if r/r != 1: # if args.samples doesn't fit into args.columns, plot all samples provided and leave the rest empty
        s=args.nsamples+(args.nsamples%args.ncol)
        r=s/args.ncol
    r=int(r)

    ax_ratio=(args.width/args.ncol)/(args.height/r)
    fig,axes=plt.subplots(nrows=r,ncols=args.ncol)

    count=0
    for ax in axes.flatten():
        sample=samples[count]
        count+=1
        # kernel density
        data=(aeg[aeg['sampleID'].str.contains(sample)]).reset_index(drop=True)
        kde=KernelDensity(bandwidth=0.04,metric='haversine',kernel='gaussian',algorithm='ball_tree')
        Xtrain=np.vstack([data.ypred,data.xpred]).T
        Xtrain=np.radians(Xtrain)
        xgrid=np.linspace(min(data.xpred)-10,max(data.xpred)+10,(int((max(data.xpred)-min(data.xpred)))*10))
        ygrid=np.linspace(min(data.ypred)-10,max(data.ypred)+10,(int((max(data.ypred)-min(data.ypred)))*10))
        Xgrid,Ygrid=np.meshgrid(xgrid,ygrid)
        kde.fit(Xtrain)
        xy=np.vstack([Ygrid.ravel(),Xgrid.ravel()]).T
        xy=np.radians(xy)
        Z=np.exp(kde.score_samples(xy))
        zed=np.sort(Z)
        c1=np.cumsum(zed)
        y_interp=sp.interpolate.interp1d(c1,zed)
        ex=np.linspace(min(zed),max(zed),len(np.ndarray.flatten(Xgrid+Ygrid)))
        why=y_interp(zed)
        vals=list(zip(ex,why))
        quants=np.quantile(why,[.05,.5,.9])
        levels=[]
        for q in quants:
            close=get_closest(why,q)
            if vals[close][0] not in levels:
                levels.append(vals[close][0])
        Z=Z.reshape(Xgrid.shape)
        # plot params
        if ((max(data.xpred)+10)-(min(data.xpred)-10))>((max(data.ypred)+10)-(min(data.ypred)-10)):
            xmin,xmax=min(data.xpred)-10,max(data.xpred)+10
            ax.set_xlim(xmin,xmax)
            width=(max(data.xpred)+10)-(min(data.xpred)-10)
            height=width/ax_ratio
            center=np.mean([max(data.ypred)+10,min(data.ypred)-10])
            ymin,ymax=center-(height/2),center+height/2
            ax.set_ylim(ymin,ymax)
            ax.set_aspect('equal')
        elif ((max(data.ypred)+10)-(min(data.ypred)-10))>((max(data.xpred)+10)-(min(data.xpred)-10)):
            ymin,ymax=min(data.ypred)-10,max(data.ypred)+10
            ax.set_ylim(ymin,ymax)
            height=(max(data.ypred)+10)-(min(data.ypred)-10)
            width=ax_ratio*height
            center=np.mean([max(data.xpred)+10,min(data.xpred)-10])
            xmin,xmax=center-(width/2),center+(width/2)
            ax.set_xlim(xmin,xmax)
            ax.set_aspect('equal')
        ax.patch.set_facecolor('#ffffff')
        # plot map
        plot_ax=ax.inset_axes([0,0,1,1])
        plot_ax.set_xlim(xmin,xmax)
        plot_ax.set_ylim(ymin,ymax)
        if args.basemap:
            itm=[]
            for group in mapp:
                for item in mapp[group]:
                    if any(xmin<xcoord<xmax for xcoord in mapp[group][item][0]) or any(ymin<ycoord<ymax for ycoord in mapp[group][item][1]):
                        ax.plot(mapp[group][item][0],mapp[group][item][1],'#ffffff',lw=.15)
                        ax.fill(mapp[group][item][0],mapp[group][item][1],'#b0b0b0')
        [s.set_visible(False) for s in ax.spines.values()]
        ax.get_xaxis().set_visible(False)
        ax.get_yaxis().set_visible(False)
        # plot locs
        contour=plot_ax.contour(Xgrid,Ygrid,Z,levels=levels,colors='k')
        fmt={}
        strs=['0.95', '0.5', '0.1']
        for l,s in zip(contour.levels,strs):
            fmt[l]=s
        plot_ax.clabel(contour,contour.levels,inline=True,fmt=fmt,fontsize='small')
        plot_ax.scatter(data.xpred,data.ypred,s=1,color='#000000',label='Predicted Locations')
        plot_ax.scatter(data.x[0],data.y[0],s=40,color='#FF0000',label='Sample Location')
        if args.training_samples:
            plot_ax.scatter(ts_x,ts_y,s=10,color='#1e90ff',label='Training Locations')
        plot_ax.patch.set_alpha(0)
        plot_ax.set_title(sample,fontsize='small')
        handles,labels=plot_ax.get_legend_handles_labels()
    loc='lower center'
    col=len(handles)
    #    fig.subplots_adjust(bottom=.01)
    lgt=fig.legend(handles,labels,loc=loc,ncol=col,fontsize='small')
    fig.set_size_inches(args.width,args.height)
    plt.savefig(args.out+'.pdf',format='pdf',bbox_inches='tight')

File: scripts/plot_locator.R
#plot output for one individual from a Locator run
suppressMessages(suppressWarnings(require(data.table)))
suppressMessages(suppressWarnings(require(scales)))
suppressMessages(suppressWarnings(require(raster)))
suppressMessages(suppressWarnings(require(sp)))
suppressMessages(suppressWarnings(require(MASS)))
suppressMessages(suppressWarnings(require(rgeos)))
suppressMessages(suppressWarnings(require(plyr)))
suppressMessages(suppressWarnings(require(progress)))
suppressMessages(suppressWarnings(require(argparse)))
suppressMessages(suppressWarnings(require(ggplot2)))

parser <- argparse::ArgumentParser(description="Plot summary of a set of locator predictions.")
parser$add_argument('--infile',help="path to folder with .predlocs files")
parser$add_argument('--sample_data',help="path to sample_data file (should be WGS1984 x / y if map=TRUE.")
parser$add_argument('--out',help="path to output (will be appended with _typeofplot.pdf)")
parser$add_argument('--width',default=5,type="double",help="width in inches of the output map. default = 5")
parser$add_argument('--height',default=4,type="double",help="height in inches of the output map. default = 4")
parser$add_argument('--samples',default=NULL,type="character",help="samples IDs to plot, separated by commas. e.g. sample1,sample2,sample3. No spaces. default = NULL")
parser$add_argument('--nsamples',default=9,help="if no --samples argument is provided, --nsamples random samples will be plotted. default = 9")
parser$add_argument('--ncol',default=3,type="integer",help="number of columns for multipanel plots (should evenly divide --nsamples). default = 3")
parser$add_argument('--error',default=FALSE,action="store_true",help="calculate error and plot summary? requires known locations for all samples. T / F. default = F")
parser$add_argument('--legend_position',default="bottom",help="legend position for summary plots if --error is True. Options:'bottom','right'. default = bottom")
parser$add_argument('--map',default="T",type="character",help="plot basemap? default = T")
parser$add_argument('--longlat',default=FALSE,action="store_true",help="set to TRUE if coordinates are x and y in decimal degrees for error in kilometers. default: FALSE. ")
parser$add_argument('--haploid',default=FALSE,action="store_true",help="set to TRUE if predictions are from locator_phased.py. Predictions will be plotted for each haploid chromosome separately. default: FALSE.")
parser$add_argument('--centroid_method',default="kd",help="Method for summarizing window/bootstrap predictions. Options 'gc' (take the centroid of window predictions with rgeos::gCentroid() ) or 'kd' (take the location of maximum density after kernal density estimation with mass::kde( )). default: kd")
args <- parser$parse_args()

infile <- args$infile
sample_data <- args$sample_data
out <- args$out
width <- args$width
height <- args$height
ncol <- args$ncol
dropout <- args$dropout
error <- args$error
samples <- args$samples
usemap <- args$map
haploid <- args$haploid
nsamples <- args$nsamples
centroid_method <- args$centroid_method

# infile <- "~/Downloads/locator/bootstraps/"
# sample_data <- "~/Downloads/locator/city_2.txt"
# out <- "~/Desktop/locator_plot_test"
# width <- 5
# height <- 4
# samples <- NULL
# nsamples<- 9
# ncol <- 3
# usemap <- T
# haploid <- F

# load("~/locator/data/cntrymap.Rdata")

kdepred <- function(xcoords,ycoords){
  try({
    density <- kde2d(xcoords,ycoords,n=500)
    max_index <- which(density[[3]] == max(density[[3]]), arr.ind = TRUE)
    kd_x <- density[[1]][max_index[1]]
    kd_y  <- density[[2]][max_index[2]]
    return(data.frame(kd_x,kd_y))
  },{
    kd_x <- mean(xcoords)
    kd_y <- mean(ycoords)
    return(data.frame(kd_x,kd_y))
  })
}

print("loading data")
if(grepl("predlocs.txt",infile)){
  pd <- fread(infile,data.table=F)
  names(pd) <- c('xpred','ypred','sampleID')
  files <- infile
} else {
  files <- list.files(infile,full.names = T)
  files <- grep("predlocs",files,value=T)
  pd <- fread(files[1],data.table=F)[0,1:3]
  for(f in files){
    a <- fread(f,data.table = F,header=T)[,1:3]
    pd <- rbind(pd,a)
  }
  names(pd) <- c('xpred','ypred','sampleID')
}

locs <- fread(sample_data,data.table=F)

if(!is.null(samples) && grepl(",",samples)){
  samples <- unlist(strsplit(samples,","))
} else if(is.null(samples)){
  samples <- sample(unique(pd$sampleID),nsamples,replace = F)
} else {
  samples <- args$samples
}

pd <- merge(pd,locs,by="sampleID")

if(error){
  print("calculating error")
  #get error for centroids and max kernel density locations
  bp <- ddply(pd,.(sampleID),function(e) {
    k <- kdepred(e$xpred,e$ypred)
    g <- as.data.frame(gCentroid(SpatialPoints(as.matrix(e[,c("xpred","ypred")]),proj4string = crs(proj4string(map)))))
    out <- unlist(c(g,k))
    names(out) <- c("gc_x","gc_y","kd_x","kd_y")
    return(out)
  })
  
  pd <- merge(pd,bp,by="sampleID")
  outsum <- pd[,c("sampleID","kd_x","kd_y","gc_x","gc_y")]
  outsum <- ddply(outsum,.(sampleID),function(e) e[1,])
  write.table(outsum,paste0(out,"_centroids.txt"),sep="\t",row.names=FALSE)
  
  plocs=as.matrix(pd[,c("kd_x","kd_y")])
  tlocs=as.matrix(pd[,c("x","y")])
  dists=sapply(1:nrow(plocs),function(e) spDistsN1(t(as.matrix(plocs[e,])),
                                                   t(as.matrix(tlocs[e,])),longlat = args$longlat))
  pd$dist_kd <- dists
  print(paste("mean kernel peak error =",mean(dists)))
  print(paste("median kernel peak error =",median(dists)))
  print(paste("90% CI for kernal peak error = ",quantile(dists,0.05),quantile(dists,0.95)))
  
  plocs=as.matrix(pd[,c("gc_x","gc_y")])
  tlocs=as.matrix(pd[,c("x","y")])
  dists=sapply(1:nrow(plocs),function(e) spDistsN1(t(as.matrix(plocs[e,])),
                                                   t(as.matrix(tlocs[e,])),longlat = args$longlat))
  pd$dist_gc <- dists
  print(paste("mean centroid error =",mean(dists)))
  print(paste("median centroid error ",median(dists)))
  print(paste("90% CI for centroid error = ",quantile(dists,0.05),quantile(dists,0.95)))
}


print("plotting")
pb <- progress_bar$new(total=length(samples))
png(paste0(out,"_windows.png"),width=width,height=height,res = 600,units = "in")
par(oma=c(0,0,0,0),mai=c(.15,.15,.15,.15),mgp=c(3,0.15,0))
if(length(samples)==1){
  layout(mat=matrix(c(1,2),byrow=T,nrow=2),heights = c(1,.5))
}
if(length(samples)==2){
  layout(mat=matrix(c(1,2,3,3),byrow=T,nrow=2),heights = c(1,.5))
} else if(length(samples)>=3){
  layout(mat=matrix(c(1:length(samples),rep(length(samples)+1,ncol)),
                    byrow=T,nrow=ceiling(length(samples)/ncol)+1),
         heights = c(rep(1,ceiling(length(samples)/ncol)),.5))
}
for(i in samples){
  print(i)
  sample <- pd[pd$sampleID==i,]
  if(usemap=="T"){
    plot(map,axes=T,cex.axis=0.5,tck=-0.03,border="white",
         xlim=c(min(na.omit(c(sample$xpred,sample$x)))-6,
                max(na.omit(c(sample$xpred,sample$x)))+6),
         ylim=c(min(na.omit(c(sample$ypred,sample$y)))-6,
                max(na.omit(c(sample$ypred,sample$y)))+6),
         col="grey",lwd=0.35)
  } else {
    plot(0,axes=T,cex.axis=0.5,tck=-0.03,
         xlim=c(min(na.omit(c(pd$xpred,pd$x)))-1,
                max(na.omit(c(pd$xpred,pd$x)))+1),
         ylim=c(min(na.omit(c(pd$ypred,pd$y)))-1,
                max(na.omit(c(pd$ypred,pd$y)))+1),
         col="white")
  }
  
  #title(paste(sample$population[1],sample$sampleID[1],sep=":"),cex.main=0.9,font.main=1)
  title(sample$sampleID[1],cex.main=0.8,font.main=1)
  box(lwd=1)
  pts <- SpatialPoints(as.matrix(data.frame(sample$xpred,sample$ypred)))
  try({
    kd <- kde2d(sample$xpred,sample$ypred,n = 80,
                lims = c(min(na.omit(c(sample$xpred,sample$x)))-15,
                         max(na.omit(c(sample$xpred,sample$x)))+15,
                         min(na.omit(c(sample$ypred,sample$y))-15),
                         max(na.omit(c(sample$ypred,sample$y)))+15))
    prob <- c(.95,.5,.1) #via https://stackoverflow.com/questions/16225530/contours-of-percentiles-on-level-plot
    dx <- diff(kd$x[1:2])
    dy <- diff(kd$y[1:2])
    sz <- sort(kd$z)
    c1 <- cumsum(sz) * dx * dy
    levels <- sapply(prob, function(x) {
      approx(c1, sz, xout = 1 - x)$y
    })
    levels <- levels[!is.na(levels)]
  },silent=TRUE)
  points(x=locs$x,y=locs$y,col="dodgerblue3",pch=1,cex=0.5,lwd=0.5)
  points(pts,pch=16,cex=0.4,col=alpha("black",0.7))
  try({
    contour(kd,levels=levels,drawlabels=T,labels=prob,add=T,
            labcex=0.32,lwd=0.5,axes=True,vfont=c("sans serif","bold"))
  },silent=TRUE)
  points(x=sample$x[1],y=sample$y[1],col="red3",pch=1,cex=.75)
  # if(!is.null(grep("FULL",files))){
  #   points(pts[grepl("FULL",files)],col="forestgreen",pch=1,cex=.8)
  # }
  #pb$tick()
}
plot(1, type = "n", axes=FALSE, xlab="", ylab="")
legend(x="top",
       legend=c("Training Locations","Sample Location","Predicted Locations"),
       col=c("dodgerblue3","red3","black"),
       pch=16,cex=1,pt.cex=2,bty='n',horiz=T,x.intersp = 1)
dev.off()


if(error){
  pdf(paste0(out,"_summary.pdf"),width=6,height=3.25,useDingbats = F)
  if(usemap=="T"){
    if(centroid_method=="gc"){
      truelocs <- ddply(pd,.(x,y),summarize,error=mean(dist_gc))
      locsn <- ddply(locs,.(x,y),summarize,n=length(sampleID))
      truelocs <- merge(truelocs,locsn,c("x","y"))
      map <- crop(map,c(min(na.omit(c(pd$xpred,pd$x)))-10,
                        max(na.omit(c(pd$xpred,pd$x)))+10,
                        min(na.omit(c(pd$ypred,pd$y)))-10,
                        max(na.omit(c(pd$ypred,pd$y)))+10))
      print(ggplot()+coord_map(projection = "mollweide",
                               xlim=c(min(na.omit(c(pd$xpred,pd$x)))-10,
                                      max(na.omit(c(pd$xpred,pd$x)))+10),
                               ylim=c(min(na.omit(c(pd$ypred,pd$y)))-10,
                                      max(na.omit(c(pd$ypred,pd$y)))+10))+
              theme_classic()+theme(axis.title = element_blank(),
                                    legend.title = element_text(size=8),
                                    legend.text=element_text(size=6),
                                    axis.text=element_text(size=6),
                                    # legend.box = "horizontal",
                                    legend.position = args$legend_position)+
              scale_color_distiller(palette = "RdYlBu",name="Mean Error\n(km)")+
              scale_size_continuous(name="Training\nSamples")+
              geom_polygon(data=fortify(map),aes(x=long,y=lat,group=group),fill="grey",color="white",lwd=0.2)+
              geom_point(data=truelocs,aes(x=x,y=y,color=error,size=n))+
              geom_segment(data=pd,aes(x=x,y=y,xend=gc_x,yend=gc_y),lwd=0.2)+
              geom_point(data=pd,aes(x=gc_x,y=gc_y),size=0.5,shape=1))
    } else if(centroid_method=="kd") {
      truelocs <- ddply(pd,.(x,y),summarize,error=mean(dist_kd))
      locsn <- ddply(locs,.(x,y),summarize,n=length(sampleID))
      truelocs <- merge(truelocs,locsn,c("x","y"))
      map <- crop(map,c(min(na.omit(c(pd$xpred,pd$x)))-10,
                        max(na.omit(c(pd$xpred,pd$x)))+10,
                        min(na.omit(c(pd$ypred,pd$y)))-10,
                        max(na.omit(c(pd$ypred,pd$y)))+10))
      print(ggplot()+coord_map(projection = "mollweide",
                               xlim=c(min(na.omit(c(pd$xpred,pd$x)))-10,
                                      max(na.omit(c(pd$xpred,pd$x)))+10),
                               ylim=c(min(na.omit(c(pd$ypred,pd$y)))-10,
                                      max(na.omit(c(pd$ypred,pd$y)))+10))+
              theme_classic()+theme(axis.title = element_blank(),
                                    legend.title = element_text(size=8),
                                    legend.text=element_text(size=6),
                                    axis.text=element_text(size=6),
                                    # legend.box = "horizontal",
                                    legend.position = args$legend_position)+
              scale_color_distiller(palette = "RdYlBu",name="Mean Error\n(km)")+
              scale_size_continuous(name="Training\nSamples")+
              geom_polygon(data=fortify(map),aes(x=long,y=lat,group=group),fill="grey50",color="white",lwd=0.2)+
              geom_point(data=locs,aes(x=x,y=y),shape=1,color="grey30",size=0.6,stroke=0.3)+
              geom_point(data=truelocs,aes(x=x,y=y,color=error,size=n))+
              geom_segment(data=pd,aes(x=x,y=y,xend=kd_x,yend=kd_y),lwd=0.2)+
              geom_point(data=pd,aes(x=kd_x,y=kd_y),size=0.5,shape=1))
    }
  } else {
    print(ggplot()+
            theme_classic()+theme(axis.title = element_blank(),
                                  legend.title = element_text(size=8),
                                  legend.text=element_text(size=6),
                                  axis.text=element_text(size=6),
                                  # legend.box = "horizontal",
                                  legend.position = args$legend_position)+
            scale_color_distiller(palette = "RdYlBu",name="Mean Error")+
            scale_size_continuous(name="Training\nSamples")+
            #geom_polygon(data=fortify(map),aes(x=long,y=lat,group=group),fill="grey",color="white",lwd=0.2)+
            geom_point(data=locs,aes(x=x,y=y),shape=1,color="grey50",size=0.6,stroke=0.3)+
            geom_point(data=truelocs,aes(x=x,y=y,color=error,size=n))+
            geom_segment(data=pd,aes(x=x,y=y,xend=gc_x,yend=gc_y),lwd=0.2)+
            geom_point(data=pd,aes(x=gc_x,y=gc_y),size=0.5,shape=1))
  }
 
  dev.off()
  
  #pd$dist <- apply(pd[,2:5],1,function(e) spDistsN1(matrix(e[1:2],ncol=2),matrix(e[3:4],ncol=2),longlat = TRUE))
  pdf(paste0(out,"_error_histogram.pdf"),width=3,height=2.5)
  print(ggplot(data=pd,aes(x=dist_gc))+
    theme_classic()+theme(axis.text=element_text(size=6),axis.title=element_text(size=8))+
    xlab("Test Error (km)")+ylab("n samples")+
    #scale_x_log10()+
    geom_histogram())
  dev.off()
}

  